### PR TITLE
Introduce BetaGeoNBD Random Variable

### PR DESCRIPTION
**Pull request by PabloRoque**
## Description

Context: Prepares the introduction of time-invariant covariates in (M)BG/NBD models

* Introduces BetaGeoNBDRV contributing to #527
* `distribution_new_customer | distribution_new_customer_dropout` now accept data as `Optional`. This is in line with other distributions, contributing to the standardization
* Adds `distribution_new_customer_recency_frequency` in line with standardization
* Overrides `ModifiedBetaGeoModel.distribution_new_customer`. Note that we cannot use `super().distribution_new_customer` until a new distribution block is added for `ModifiedBetaGeoRV`

## Related Issue

- [ ] Closes #
- [x] Related to #527 #128
- [x] Related to #1230 

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Modules affected

- [ ] MMM
- [x] CLV

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pymc-marketing--1317.org.readthedocs.build/en/1317/

### CODE REVIEW

File: pymc_marketing/clv/distributions.py [hunk 1]
```diff
@@ -662,3 +587,152 @@ def logp_customer_died(t_x_i, x_i, alpha_i, beta_i, gamma_i, delta_i, T_i):
             delta > 0,
             msg="alpha > 0, beta > 0, gamma > 0, delta > 0",
         )
+
+
+class BGNBDRV(RandomVariable):
```
Comment (ColtAllen) [inline pymc_marketing/clv/distributions.py:592]: Can we rename to `BetaGeoNBD`?

File: pymc_marketing/clv/models/beta_geo.py [hunk 2]
```diff
@@ -196,53 +198,66 @@ def build_model(self) -> None:  # type: ignore[override]
...
+            #         x=self.data["frequency"],
+            #         t_x=self.data["recency"],
+            #         a=a,
+            #         b=b,
+            #         alpha=alpha,
+            #         r=r,
+            #         T=self.data["T"],
+            #     ),
+            # )
+
```
Comment (ColtAllen) [inline pymc_marketing/clv/models/beta_geo.py:260]: Why is this commented-out code being added?

File: tests/clv/models/test_beta_geo.py [hunk 3]
```diff
@@ -213,6 +214,76 @@ def test_numerically_stable_logp(
...
+            rtol=0.5,
+        )
+        np.testing.assert_allclose(customer_dropout.mean(), mu_mean, rtol=0.5)
+        np.testing.assert_allclose(customer_dropout.std(), mu_std, rtol=0.5)
+
+        np.testing.assert_allclose(customer_rec.mean(), ref_rec.mean(), rtol=0.5)
+        np.testing.assert_allclose(customer_rec.std(), ref_rec.std(), rtol=0.5)
+
+        np.testing.assert_allclose(customer_freq.mean(), ref_freq.mean(), rtol=0.5)
+        np.testing.assert_allclose(customer_freq.std(), ref_freq.std(), rtol=0.5)
```
Comment (ColtAllen) [inline tests/clv/models/test_beta_geo.py:285]: The new distribution block will need to be added to `_distribution_new_customers`. Here's an [example](https://github.com/pymc-labs/pymc-marketing/blob/main/pymc_marketing/clv/models/pareto_nbd.py#L947) of what would look like.

`_distribution_new_customers` will also need to be overridden for `ModifiedBetaGeoModel` if a block for that model isn't added before the next release.

File: tests/clv/test_distributions.py [hunk 4]
```diff
@@ -443,3 +444,72 @@ def test_beta_geo_beta_binom_sample_prior(
 
         np.testing.assert_allclose(lt_frequency.mean(), recency.mean(), rtol=0.84)
         np.testing.assert_allclose(lt_recency.mean(), frequency.mean(), rtol=0.84)
+
+
+class TestBGNBD:
+    def test_logp_matches_excel(self):
```
Comment (ColtAllen) [inline tests/clv/test_distributions.py:450]: Add a docstring citing where in the literature this is referenced.

Also, why aren't any of the other tests used in other distribution blocks copied over?

**Review by ColtAllen — CHANGES_REQUESTED** [review]
It's a good start, but it looks like a lot of changes were copied over from https://github.com/pymc-labs/pymc-marketing/pull/1269. This will create merge conflicts and/or require a rebase to preserve @ricardoV94's contributions in that PR.

Can you revert the changes to the other distribution blocks? I would prefer to merge https://github.com/pymc-labs/pymc-marketing/pull/1269 before proceeding with this one.

File: pymc_marketing/clv/models/beta_geo.py (hunk 2 above)
Comment (PabloRoque) [reply pymc_marketing/clv/models/beta_geo.py:260]: Sorry, I was checking that the tests gave the same outputs. 

Removed the commented block

File: pymc_marketing/clv/distributions.py (hunk 1 above)
Comment (PabloRoque) [reply pymc_marketing/clv/distributions.py]: You got it

**Comment by ColtAllen** [comment]
> As explained above I cannot revert, because basically nothing runs on my setup if reverted.

Here's a [workaround](https://github.com/pymc-labs/pymc-marketing/issues/1305#issuecomment-2557501267) for MacOS.

File: tests/clv/test_distributions.py (hunk 4 above)
Comment (PabloRoque) [reply tests/clv/test_distributions.py:527]: Added

File: pymc_marketing/clv/distributions.py [hunk 5]
```diff
@@ -662,3 +668,169 @@ def logp_customer_died(t_x_i, x_i, alpha_i, beta_i, gamma_i, delta_i, T_i):
...
         )
+
+
+class BetaGeoNBDRV(RandomVariable):
+    name = "bg_nbd"
+    ndim_supp = 1
+    ndims_params = [0, 0, 0, 0, 0]
+
+    dtype = "floatX"
+    _print_name = ("BetaGeoNBD", "\\operatorname{BetaGeoNBD}")
```
Comment (wd60622) [inline pymc_marketing/clv/distributions.py:679]: This will change with pymc dependency update #1269
I am unclear on that timeline. 
Should we wait on that PR first? What are the thoughts?

Comment (PabloRoque) [reply pymc_marketing/clv/distributions.py:679]: I really have no preferences, it depends whether #1269 is merged before this PR or the other way around.

It was originally committed with the new signature param. Reverted it so all the RVs follow the same format.

File: tests/clv/models/test_beta_geo.py (hunk 3 above)
Comment (PabloRoque) [reply tests/clv/models/test_beta_geo.py]: > _distribution_new_customers will also need to be overridden for ModifiedBetaGeoModel if a block for that model isn't added before the next release.

This has now been overridden, although still following new naming conventions

File: pymc_marketing/clv/distributions.py (hunk 5 above)
Comment (PabloRoque) [reply pymc_marketing/clv/distributions.py]: Since #1269 has been merged, I introduced back the signature param

File: pymc_marketing/clv/models/beta_geo.py [hunk 6]
```diff
@@ -456,6 +425,36 @@ def expected_probability_alive(
...
+            alpha=alpha.values[..., None]
+            if "customer_id" not in alpha.dims
+            else alpha.values,
+            T=T.values,
+        )
+        values = np.vstack((t_x.values, x.values)).T
+        # TODO: Instead of compiling this function everytime this method is called
+        #  we could compile it once (with mutable inputs) and cache it for reuse with new inputs.
+        loglike = pm.logp(bg_bnd_dist, values).eval()
+        return xarray.DataArray(data=loglike, dims=("chain", "draw", "customer_id"))
```
Comment (ColtAllen) [inline pymc_marketing/clv/models/beta_geo.py:456]: This function can be removed because it isn't used in any of the predictive methods.

File: tests/clv/models/test_beta_geo.py [hunk 7]
```diff
@@ -174,6 +176,122 @@ def test_customer_id_duplicate(self):
...
+            lam_mean,
+            rtol=0.5,
+        )
+        np.testing.assert_allclose(
+            customer_purchase_rate.std(),
+            lam_std,
+            rtol=0.5,
+        )
+        # np.testing.assert_allclose(customer_dropout.mean(), mu_mean, rtol=0.5)
+        # np.testing.assert_allclose(customer_dropout.std(), mu_std, rtol=0.5)
```
Comment (ColtAllen) [inline tests/clv/models/test_beta_geo.py:287]: Why are these code lines commented out? The dropout distribution in particular isn't being tested (you can also rename it from `mu_` to `p_` for naming consistency.)

File: tests/clv/test_distributions.py [hunk 8]
```diff
@@ -443,3 +445,201 @@ def test_beta_geo_beta_binom_sample_prior(
...
+                np.full((1), 11.67),
+                12,
+            ),
+        ],
+    )
+    def test_bg_nbd(self, value, r, alpha, a, b, T):
+        # weights = np.ones_like(value, dtype=int)
+
+        def lifetimes_wrapper(
+            r, alpha, a, b, freq, rec, T, weights=np.array(1), penalizer_coef=0.0
```
Comment (ColtAllen) [inline tests/clv/test_distributions.py:500]: The commented-out `weights` variable can be removed if the corresponding lifetimes parameter isn't needed.

**Review by ColtAllen — CHANGES_REQUESTED** [review]
Looks good so far! Just need to revise the `sim_data` function, docstrings, and a few other things.

File: pymc_marketing/clv/distributions.py [hunk 9]
```diff
@@ -587,3 +593,156 @@ def logp_customer_died(t_x_i, x_i, alpha_i, beta_i, gamma_i, delta_i, T_i):
...
+                t += wait
+                n += 1
+                wait = rng.exponential(scale=1 / lam)
+
+            return np.array(
+                [
+                    t,
+                    n,
+                ],
+            )
```
[the remark this answers is not in the document]
Comment (PabloRoque) [reply pymc_marketing/clv/distributions.py]: Added

Gorgeous comment. I was sure there was something off there. Tried my best though :_D

File: pymc_marketing/clv/distributions.py [hunk 10]
```diff
@@ -587,3 +593,156 @@ def logp_customer_died(t_x_i, x_i, alpha_i, beta_i, gamma_i, delta_i, T_i):
...
+        return output
+
+
+bg_nbd = BetaGeoNBDRV()
+
+
+class BetaGeoNBD(PositiveContinuous):
+    r"""Population-level distribution class for a discrete, non-contractual, Beta-Geometric/Negative-Binomial process.
+
+    It is based on Fader, et al. in [1]_ and [2]_.
```
[the remark this answers is not in the document]
Comment (PabloRoque) [reply pymc_marketing/clv/distributions.py]: Added

File: pymc_marketing/clv/models/beta_geo.py (hunk 6 above)
Comment (PabloRoque) [reply pymc_marketing/clv/models/beta_geo.py]: Deleted

File: tests/clv/test_distributions.py (hunk 8 above)
Comment (PabloRoque) [reply tests/clv/test_distributions.py:500]: Removed

File: tests/clv/models/test_beta_geo.py (hunk 7 above)
Comment (PabloRoque) [reply tests/clv/models/test_beta_geo.py]: Fixed

File: pymc_marketing/clv/distributions.py [hunk 11]
```diff
@@ -587,3 +593,156 @@ def logp_customer_died(t_x_i, x_i, alpha_i, beta_i, gamma_i, delta_i, T_i):
...
+        where:
+        A_1 &= \frac{\Gamma(r+x)\alpha^r}{\Gamma(r)} \\
+        A_2 &= \frac{\Gamma(a+b)\Gamma(b+x)}{\Gamma(b)\Gamma(a+b+x)} \\
+        A_3 &= (\frac{1}{\alpha + T})^{r+x} \\
+        A_4 &= (\frac{a}{b+x-1})(\frac{1}{\alpha + t_x})^{r+x} \\
+
+    ========  ===============================================
+    Support   :math:`t_j >= 0` for :math:`j = 1, \dots,x`
+    Mean      :math:`\mathbb{E}[X(n) | r, \alpha, a, b] = \frac{a+b-1}{a-1} \left[ 1 - (\frac{\alpha}{\alpha + T})^r \leftidx{_2}{F}{_1}(r,b;a+b-1;\frac{t}{\alpha + t}) \right] `
+    ========  ===============================================
```
[the remark this answers is not in the document]
Comment (PabloRoque) [reply pymc_marketing/clv/distributions.py]: Had to render it a few times locally, but it should be fixed now.

File: pymc_marketing/clv/distributions.py [hunk 12]
```diff
@@ -587,3 +593,152 @@ def logp_customer_died(t_x_i, x_i, alpha_i, beta_i, gamma_i, delta_i, T_i):
...
+        def sim_data(lam, p, T):
+            t = 0  # recency
+            n = 0  # frequency
+
+            churn = 0  # BG/NBD assumes all non-repeat customers are active
+            wait = rng.exponential(scale=1 / lam)
+
+            while t + wait < T and not churn:
+                n += 1
+                churn = rng.binomial(n=1, p=p)
```
Comment (ColtAllen) [inline pymc_marketing/clv/distributions.py:642]: Ah, I made two mistakes in that comment I shared regarding this. Here's the fix:

1. Change this line to `churn = rng.random() < p`
2. Move it to before `n+=1`

File: pymc_marketing/clv/distributions.py [hunk 13]
```diff
@@ -587,3 +593,152 @@ def logp_customer_died(t_x_i, x_i, alpha_i, beta_i, gamma_i, delta_i, T_i):
...
+    .. math::
+
+        \mathbb{LL}(r, \alpha, a, b  | x, t_x, T) =
+        D_1 + D_2 + \ln(C_3 + \delta_{x>0} C_4) \text{, where:} \\
+        \begin{align}
+        D_1 &= \ln \left[ \Gamma(r+x) \right] - \ln \left[ \Gamma(r) \right] + \ln \left[ \Gamma(a+b) \right] + \ln \left[ \Gamma(b+x) \right] \\
+        D_2 &= r \ln(\alpha) - (r+x) \ln(\alpha + t_x) \\
+        C_3 &= \left(\frac{\alpha + t_x}{\alpha + T} \right)^{r+x} \\
+        C_4 &= \left(\frac{a}{b+x-1} \right) \\
+        \end{align}
```
Comment (ColtAllen) [inline pymc_marketing/clv/distributions.py:671]: Hmm, the docs preview isn't showing these changes yet. I'll check again later.

**Review by ColtAllen — CHANGES_REQUESTED** [review]
Almost there!

File: pymc_marketing/clv/distributions.py (hunk 12 above)
Comment (PabloRoque) [reply pymc_marketing/clv/distributions.py]: Done

File: pymc_marketing/clv/distributions.py (hunk 13 above)
Comment (PabloRoque) [reply pymc_marketing/clv/distributions.py:671]: I see them here from the last build: https://pymc-marketing--1317.org.readthedocs.build/en/1317/api/generated/pymc_marketing.clv.distributions.BetaGeoNBD.html

Let me know if you find anything funny

**Review by ColtAllen — APPROVED** [review]
Thanks for adding this! This will allow use of the `utils.plot_purchase_pmf` function for `BetaGeoModel`, as well as permit `arviz` metrics like WAIC for out-of-sample predictive power.

If you wish to create an equivalent RV for `ModifiedBetaGeoModel`, you can probably just inherit from these classes and override the `rng_fn` and `logp` methods.

**Comment by PabloRoque** [comment]
I truly appreciate the time you took to go over it. 
Expect that PR you mention, and one introducing covars for each of the 2 models.

**Comment by PabloRoque** [comment]
@wd60622 Could you have a look at this PR? If you don't see any issues could you consider merging it?

I am preparing `ModifiedBetaGeoNBDRV` and would be good to rebase to main with this PR merged.

Thanks!

File: tests/test_mlflow.py [hunk 14]
```diff
@@ -483,7 +483,7 @@ def test_clv_fit_mcmc(model_cls, clv_data) -> None:
     run_id = run.info.run_id
     inputs, params, metrics, tags, artifacts = get_run_data(run_id)
 
-    assert inputs == []
+    assert isinstance(inputs, list)
```
Comment (wd60622) [inline tests/test_mlflow.py:486]: what was happening here?

**Review by wd60622 — APPROVED** [review]
Looks good. I just had question about the mlflow test. I'll merge still

File: tests/test_mlflow.py (hunk 14 above)
Comment (PabloRoque) [reply tests/test_mlflow.py:486]:
```
@pytest.mark.parametrize("model_cls", [BetaGeoModel])
    def test_clv_fit_mcmc(model_cls, clv_data) -> None:
        mlflow.set_experiment("pymc-marketing-test-suite-clv")
    
        sampler_config = {
            "draws": 2,
            "chains": 1,
            "tune": 1,
        }
    
        model = model_cls(data=clv_data, sampler_config=sampler_config)
        with mlflow.start_run() as run:
            model.fit()
    
        assert mlflow.active_run() is None
    
        run_id = run.info.run_id
        inputs, params, metrics, tags, artifacts = get_run_data(run_id)
    
>       assert inputs == []
E       assert [<DatasetInpu...e='sample'>]>] == []
E         Left contains one more item: <DatasetInput: dataset=<Dataset: digest='cebda4ee', name='dataset', profile=('{"features_shape": {}, "features_size": ... '"mlflow.source.type": "LOCAL"}}'), source_type='code'>, tags=[<InputTag: key='mlflow.data.context', value='sample'>]>
E         Full diff:
E           [
E         -  ,
E         +  <DatasetInput: dataset=<Dataset: digest='cebda4ee', name='dataset', profile=('{"features_shape": {}, "features_size": {}, "features_nbytes": {}, '
E         +  '"targets_shape": {"recency_frequency": [4, 2]}, "targets_size": '
E         +  '{"recency_frequency": 8}, "targets_nbytes": {"recency_frequency": 64}}'), schema=None, source=('...
E         
E         ...Full output truncated (4 lines hidden), use '-vv' to show

tests/test_mlflow.py:486: AssertionError
```

Comment (PabloRoque) [reply tests/test_mlflow.py:486]: `inputs` is not an empty array

Comment (wd60622) [reply tests/test_mlflow.py:486]: I see. thanks for the traceback
